File: NET/JNetCall.Sharp.InProc/Server/ServiceLots.cs
using System;
using System.Collections.Generic;
using System.IO;
using JNetBase.Sharp.Threads;
using JNetCall.Sharp.API;
using JNetCall.Sharp.API.IO;
using JNetCall.Sharp.Tools;

namespace JNetCall.Sharp.Server
{
    public static class ServiceLots
    {
        public static ClassHosting Create<T>() where T : new()
        {
            var protocol = new NativeHostSink();
            Register(protocol);
            return Create<T>(protocol);
        }

        private static ClassHosting Create<T>(ISendTransport protocol) where T : new()
        {
            var instance = new T();
            var pool = new ThreadExecutor();
            var host = new ClassHosting(instance, protocol, pool);
            return host;
        }

        private static readonly IList<ICaller> Lots = new List<ICaller>();

        private static void Register(ICaller lot)
        {
            Lots.Add(lot);
        }

        // ReSharper disable UnusedMember.Global
        private static byte[] Call(byte[] input)
        {
            foreach (var lot in Lots)
            {
                using var output = new MemoryStream();
                if (!lot.TryCall(input, output))
                    continue;
                return output.ToArray();
            }
            return new[] { unchecked((byte)-1) };
        }

        public static IntPtr Call(IntPtr inputPtr)
        {
            var input = Interop.ToByteArray(inputPtr);
            var output = Call(input);
            return Interop.ToPointer(output);
        }
    }
}

File: JVM/jnetcall-java-inproc/src/main/java/jnetcall/java/server/ServiceLots.java
package jnetcall.java.server;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;

import jnetbase.java.meta.Reflect;
import jnetbase.java.threads.ThreadExecutor;
import jnetcall.java.api.ICaller;
import jnetcall.java.api.io.ISendTransport;

public final class ServiceLots {

    public static <T> ClassHosting create(Class<T> serviceClass) {
    	ServiceLot protocol = new ServiceLot();
        register(protocol);
        return create(serviceClass, protocol);
    }

    private static <T> ClassHosting create(Class<T> serviceClass, ISendTransport protocol) {
        T instance = Reflect.createNew(serviceClass);
        ThreadExecutor pool = new ThreadExecutor();
        ClassHosting host = new ClassHosting(instance, protocol, pool);
        return host;
    }

    private static final List<ICaller> lots = new ArrayList<>();

    private static void register(ICaller lot) {
        lots.add(lot);
    }

    @SuppressWarnings("unused")
    public static byte[] call(byte[] input) throws Exception {
        for (ICaller lot : lots)
            try (ByteArrayOutputStream output = new ByteArrayOutputStream()) {
                if (!lot.tryCall(input, output))
                    continue;
                return output.toByteArray();
            }
        return new byte[]{-1};
    }
}